### ISSUE TITLE
Improving Play!'s speed

### ISSUE THREAD

**Issue by MoochMcGee** [jpd002/Play-#352]
Currently, Play!'s speed is abysmal compared to that of PCSX2. PCSX2 commonly runs 10x faster ingame, even with release builds, thereby making debug builds nigh-unusable.

I propose this issue as a way for us to collaborate on getting Play! to playable speeds. Yes, the pun was intended.

Here is a list of potential speed issues, and I request that at least @jpd002 respond to questions about them.

- [ ] Register allocation woes
 - [ ] Figuring out what kinds of register allocations can be used, and weigh their advantages and disadvantages.
- [ ] Incredibly slow GS performance when it comes to working with lots of draw calls.

**Comment by jpd002** [comment]
From the tests I've done a while ago, in general, emulation speed is bound by EE execution (by a large margin), then VU and then VIF.

Register allocation is an issue, but I don't think it's the biggest one. I think the lack of block linking and lack of proper speed hacks are the things that hurt the most. Enabling direct memory access (to RAM or SPR) instead of going through a memory map function would help a lot. 

GS rendering is not a bottleneck that came up often. When it was a bottleneck, it was because of a game that's uploading lots of textures. The slowness might seem related to the number of "draw calls", but most of the time it's because of the amount of vertices the VU code has to process. It's a bit different on mobile, but it's still not the biggest bottleneck.

**Comment by MoochMcGee** [comment]
What do you mean by "lack of proper speed hacks"?

**Comment by jpd002** [comment]
Detecting idle loops and making the emulator go faster when they are being executed.

**Comment by yorkiesandskittles** [comment]
Compared to PCSX2, Play! is still somewhat early on in development  Most emulators have the initial goal of basically making sure everything works (as few completely broken games as possible), after which speed can be focused on and improved. Better to have a slow emulator that runs games well and accurately, rather than a fast one with graphical glitching or physics inaccuracies all over the place.

Of course, speed **will** see improvements, just not a lot of them or many big ones right now.

**Comment by MoochMcGee** [comment]
@yorkiesandskittles Yeah, I know that, dude, I just want to be able to debug shit without waiting half a year for the game to load.

**Comment by jpd002** [comment]
To improve speed in debug mode, it might be possible to change the way breakpoints are handled. I usually get around by using save states or using the "ReleaseWithDebugger" build configuration.

**Comment by MoochMcGee** [comment]
Weird, because that build configuration doesn't have a debugger.

**Comment by jpd002** [comment]
It's there under the "Debug" menu. You can also dump frames and use the Frame Debugger in "ReleaseWithDebugger" config.

**Comment by MoochMcGee** [comment]
That's not there in that config.

**Comment by MoochMcGee** [comment]
Okay, what the hell? When I hit rebuild all on ReleaseWithDebugger, suddenly the Debug menu is there! WTF?!
